src/ctap/pin_protocol.rs

```diff
@@ -0,0 +1,460 @@
+            .ok_or(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER)?;
+        *array_ref!(iv_block, 0, 16)
+    } else {
+        [0u8; 16]
+    };
+    let mut blocks = Vec::with_capacity(block_len);
+    for block in block_iter {
+        blocks.push(*array_ref!(block, 0, 16));
+    }
+    let aes_enc_key = crypto::aes256::EncryptionKey::new(key);
+    let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+    cbc_decrypt(&aes_dec_key, iv, &mut blocks);
+    Ok(blocks.iter().flatten().cloned().collect::<Vec<u8>>())
+}
+
+fn verify_v1(key: &[u8], message: &[u8], signature: &[u8]) -> Result<(), Ctap2StatusCode> {
+    if signature.len() != 16 {
+        return Err(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER);
+    }
+    if verify_hmac_256_first_128bits::<Sha256>(key, message, array_ref![signature, 0, 16]) {
+        Ok(())
+    } else {
+        Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID)
+    }
+}
+
+pub struct SharedSecretV1 {
+    common_secret: [u8; 32],
+}
+
+impl SharedSecretV1 {
+    /// Creates a new shared secret from the handshake result.
+    fn new(handshake: [u8; 32]) -> SharedSecretV1 {
+        SharedSecretV1 {
+            common_secret: Sha256::hash(&handshake),
+        }
+    }
+
+    /// Creates a new shared secret for testing.
+    #[cfg(test)]
+    pub fn new_test(hash: [u8; 32]) -> SharedSecretV1 {
+        SharedSecretV1 {
+            common_secret: hash,
+        }
+    }
+}
+
+impl SharedSecret for SharedSecretV1 {
+    fn encrypt(
+        &mut self,
+        rng: &mut dyn Rng256,
+        plaintext: &[u8],
+    ) -> Result<Vec<u8>, Ctap2StatusCode> {
+        encrypt(rng, &self.common_secret, plaintext, false)
+    }
+
+    fn decrypt(&mut self, ciphertext: &[u8]) -> Result<Vec<u8>, Ctap2StatusCode> {
+        decrypt(&self.common_secret, ciphertext, false)
+    }
+
+    fn verify(&mut self, message: &[u8], signature: &[u8]) -> Result<(), Ctap2StatusCode> {
+        verify_v1(&self.common_secret, message, signature)
+    }
+}
+
+fn verify_v2(key: &[u8], message: &[u8], signature: &[u8]) -> Result<(), Ctap2StatusCode> {
+    if signature.len() != 32 {
+        return Err(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER);
+    }
+    if verify_hmac_256::<Sha256>(key, message, array_ref![signature, 0, 32]) {
+        Ok(())
+    } else {
+        Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID)
+    }
+}
+
+pub struct SharedSecretV2 {
+    handshake: [u8; 32],
+    aes_key: Option<[u8; 32]>,
+    hmac_key: Option<[u8; 32]>,
+}
+
+impl SharedSecretV2 {
+    /// Creates a new shared secret from the handshake result.
+    fn new(handshake: [u8; 32]) -> SharedSecretV2 {
+        SharedSecretV2 {
+            handshake,
+            aes_key: None,
+            hmac_key: None,
+        }
+    }
+
+    fn get_aes_key(&mut self) -> &[u8; 32] {
+        let handshake = &self.handshake;
+        self.aes_key
+            .get_or_insert_with(|| hkdf_empty_salt_256::<Sha256>(handshake, b"CTAP2 AES key"))
```

Comment:
Now I see where all these `mut` come from :)
Can't this be done in the `new` method instead?
Semantically a `get_*` is a RO accessor so having a mutable object is not really intuitive. And this would remove pretty much all the `mut` at once.

Reply:
We could remove the options around `aes_key` and `hmac_key` and initialize both right away. Downside is that we are slower in case we don't need both secrets, i.e. if there is an error before we even need the second crypto operation.
The lazy initialization is an optimization, up to you if we actually want it or not.

Reply:
Another option would be to always compute and not cache the secrets as objet properties.
Not sure in "real life scenarios" which one is worse than the other though.

Reply:
I'd definitely not want to re-compute the secrets, as they are used multiple times, i.e. if you encrypt and decrypt in one command. In "real life", we usually need both AES and HMAC key. So for me, it's between the implementation detail choices inside `SharedSecret`, if we want to compute lazy or not.

Reply:
Note that we could also use `RefCell` ([example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a49775d43492e69832c7cb84d388d48f)) to get RO-like interface to the cost of some runtime borrow check. So it's at least 3 options:
1. Recompute as needed. Pro: RO-like. Con: Large CPU usage
2. Use RefCell. Pro: RO-like. Con: Small CPU usage, Returns Ref<T> instead of &T (same usage though)
3. Use `&mut`. Pro: Returns &T. Con: RW-like, Smaller CPU usage.

Reply:
I don't see the difference between 2 and 3? I thought this is about implementing `SharedSecretV2` as

```rust
pub struct SharedSecretV2 {
    aes_key: [u8; 32],
    hmac_key: [u8; 32],
}
```

instead of 

```rust
pub struct SharedSecretV2 {
    handshake: [u8; 32],
    aes_key: Option<[u8; 32]>,
    hmac_key: Option<[u8; 32]>,
}
```

?

Reply:
We can actually save a call to `borrow` in the reuse case:

```rust
let mut cache = self.aes_key.borrow();
if cache.is_none() {
    drop(cache);  // Important otherwise panics at runtime.
    *self.aes_key.borrow_mut() = ...;
    cache = self.aes_key.borrow();
}
Ref::map(cache, ...)
```

Reply:
The current version is hopefully on a sweet spot we can agree on:
- no `mut`
- no recomputation of crypto, except for a rare conversion from encryption key to decryption key, if we do multiple decrypts in one command (very rare)
- very little RAM overhead (32B for V1, no overhead for V2, making them the same size)
- we only compute unused secrets if we fail early, which shouldn't affect usability